Pause session till all torrents are fully started

sledgehammer999 opened the pull request:
There were many user reports that complained that qbt started downloading over their data when they had forgotten to plugin the drive the torrents were in. Or forgotten to mount the partition/share etc.

I was confused because I personally had implemented a measure against this. I was even more confused because I couldn't reproduce it. Until some days ago. After a lot of trial and error it turns out that during startup libtorrent produces a ton of alerts. A lot more than 1000 at a time. And these alerts ended up throwing out of the queue the `fastresume_rejected_alert` alerts, thus my counter measure was never activated. A longer discussion here: https://github.com/arvidn/libtorrent/issues/3045

Naturally as I had delved into this I discovered other inefficiencies. There's a timing issue where the torrent might start downloading and writing data to disk before we have the chance to pause it.
I opted to have the session paused untill all torrents were loaded and handled.

To my experience, the 3 alerts that interest us are received in this order:
`add_torrent_alert` -> `fastresume_rejected_alert` -> `torrent_added_alert`

About the "Temp solution" commit. This is temporary fix in order to publish the PR and have feedback from you. Relevant bug report: https://github.com/arvidn/libtorrent/issues/3071

IMO, it's better to see each commit by itself.

Caveats:
1. I didn't thoroughly test the PR. Please help by doing runs yourselves too.
2. The RC_1_0 code path wasn't even tested if it compiles. I will do it as final step before merge.

wesleybl commented:
@sledgehammer999  if I remove the disk where the torrent was downloaded and start the qbittorrent what will happen with the percentage completed? Will keep 100% but will indicate error?

glassez commented:
@sledgehammer999, well, great job!
Although, it requires more changes.
1. We need to return the correct components loading sequence (startupTorrents should again become a part of the initialization of the session, and not be called from outside).
2. Torrent-manipulation actions (e.g. add torrent, delete torrent) should be unavailable (or maybe queued) until session is fully initialized (i.e. resumed all torrents).
3. We no need more two signals (torrentAdded/torrentNew) since it will indicate the same event.
4. Some GUI components should be disabled until session is fully initialized (or unavailable in some other way, e.g. we can show no GUI or show progress dialog).

glassez commented on src/gui/categoryfiltermodel.cpp:
```diff
@@ -181,6 +181,12 @@ CategoryFilterModel::CategoryFilterModel(QObject *parent)
     connect(session, &Session::torrentCategoryChanged, this, &CategoryFilterModel::torrentCategoryChanged);
     connect(session, &Session::subcategoriesSupportChanged, this, &CategoryFilterModel::subcategoriesSupportChanged);
+    connect(session, &Session::finishedStartingUp, this,
+            [this, session]()
```
I would move it on the previous line in all similar cases of this PR.

sledgehammer999 replied:
Do you mean the lambda? Have it as a one liner?

sledgehammer999 replied:
On 2nd thought, you probably mean only the capture clause and parameter list.

glassez replied:
>On 2nd thought, you probably mean only the capture clause and parameter list.

Yes.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -3808,11 +3786,14 @@ void Session::startUpTorrents()
         QByteArray data;
     } TorrentResumeData;
 
+    m_nativeSession->pause();
```
It would be better if we can create session in paused state.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -4009,6 +3990,12 @@ void Session::handleAlert(libt::alert *a)
             handleAddTorrentAlert(static_cast<libt::add_torrent_alert*>(a));
             break;
+        case libt::torrent_added_alert::alert_type:
+            if (m_startupInProgress)
```
Please don't break the current style. Just add `handleTorrentAddedAlert()` method.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -4009,6 +3990,12 @@ void Session::handleAlert(libt::alert *a)
+            if (m_startupInProgress)
+                --m_startedCount;
+            if (m_startedCount == 0 && m_startupInProgress)
+                handleStartUpFinish();
```
handleStartUpFinished

glassez commented on src/base/bittorrent/session.h:
```diff
@@ -538,6 +538,7 @@ namespace BitTorrent
         void subcategoriesSupportChanged();
         void tagAdded(const QString &tag);
         void tagRemoved(const QString &tag);
+        void finishedStartingUp();
```
startupFinished

Chocobo1 commented on src/gui/categoryfiltermodel.cpp:
```diff
@@ -181,6 +181,12 @@ CategoryFilterModel::CategoryFilterModel(QObject *parent)
+    connect(session, &Session::finishedStartingUp, this,
+            [this, session]()
+    {
+        for (const auto &handle : session->torrents().values())
```
add `qAsConst()`?
or use iterator to loop, this avoids redundant traverse over the container, although I like the use of values() :P

sledgehammer999 replied:
Sure. I am still confused about those macros. Eventually I'll get the hang of them.

Chocobo1 commented:
>Torrent-manipulation actions (e.g. add torrent, delete torrent) should be unavailable (or maybe queued) until session is fully initialized (i.e. resumed all torrents).

I agree with the idea, making it unavailable seems simpler.

>Some GUI components should be disabled until session is fully initialized (or unavailable in some other way, e.g. we can show no GUI or show progress dialog).

Since this is an interactive program, displaying no GUI would be a bad choice.
*If feasible*, I would choose to ready the mainwindow as early as possible (modern browsers startup time is impressive!) and let user know program is still initializing: could be some text in statusbar, or display the torrent jobs initially as "not yet loaded" and transition to other ready state (paused or downloading) when it is loaded.

glassez commented:
>or display the torrent jobs initially as "not yet loaded" and transition to other ready state (paused or downloading) when it is loaded.

Session doesn't emit signals for resumed torrents in changed code, only one startupFinished signal, and I really like this idea.

glassez commented:
>I agree with the idea, making it unavailable seems simpler.

We still need queuing at least for "add torrent" action since the application can start with torrents specified in command line. Of course, we can implement it outside the Session.

sledgehammer999 commented:
I addressed some of the comments in my 3rd commit

I pushed 2 new commits. Now the initial loading of the torrents is done in a separate thread. However, for me, the mainwindow still takes quite a few seconds to draw its contents. The initial blank/white window is drawn quickly. The widgets are taking quite a few seconds.
The 2 new commits were written rather hastily with not enough testing. So rip them apart in the review.

Chocobo1 commented:
>Now the initial loading of the torrents is done in a separate thread. However, for me, the mainwindow still takes quite a few seconds to draw its contents. 

Haven't tried this PR yet, (for you) does it still take seconds to draw the GUI even when there are no torrents?

wesleybl commented:
@sledgehammer999 I would like to test this PR in Windows. If I do:

```bash
./configure
make && make install
qbittorrent
```

on Linux and copying the files to Windows will work?

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1349,11 +1349,7 @@ void Session::configure(libtorrent::settings_pack &settingsPack)
-    // Ignore -1 values because we don't want to set a max int message queue
-    settingsPack.set_int(libt::settings_pack::alert_queue_size, std::max(1000,
-        10 * std::max(maxActiveTorrents() * 2, maxConnections())));
+    settingsPack.set_int(libt::settings_pack::alert_queue_size, std::numeric_limits<int>::max());
```
From my experience, pushing to the max limit is rarely a good idea (you reported a libtorrent bug uncovered from this).
Personally I would use a smaller limit such as INT_MAX/2 or (INT_MAX - 100) to avoid hitting the edge case, final decision is up to you.

wesleybl commented:
>on Linux and copying the files to Windows will work?

In fact, I've seen it's more complicated than that:

https://github.com/qbittorrent/qBittorrent/wiki/Compiling-with-MSVC-2015-%28static-linkage%29

I'll try it on Linux.

sledgehammer999 commented:
@wesleybl no linux binaries can't work on Windows and vice versa.
In any case I have prepared a Windows build of this PR. Use at your caution: https://builds.shiki.hu/temp/qbittorrent_4.2.0alpha_x64_PR_9007.7z

wesleybl commented:
@sledgehammer999 

https://builds.shiki.hu/temp/qbittorrent_4.2.0alpha_x64_PR_9007.7z

It didn't work on my Windows 8.1 32 bit. Is there a 32 bit version?

sledgehammer999 commented:
@wesleybl here's a 32-bit build: https://builds.shiki.hu/temp/qbittorrent_4.2.0alpha_x32_PR_9007.7z

glassez commented on src/app/application.cpp:
```diff
@@ -504,53 +504,19 @@ int Application::exec(const QStringList &params)
+    m_paramsQueue = params + m_paramsQueue;
 
     // Now UI is ready to process signals from Session
     BitTorrent::Session::instance()->startUpTorrents();
```
Is it still impossible to move it back to Session constructor?

sledgehammer999 replied:
We don't want that. We have to have created the mainwindow (with its children) before we start emitting signals from the Session.

glassez replied:
But in your current implementation you don't emit any signals from startUpTorrents directly. They will be emitted wgen message loop is started, i.e. when main window is created.

sledgehammer999 replied:
The commit that introduced this was 21a72c651f4439352942bec3573f0325b99765e9
The explanation in its commit message seems still relevant. If I am missing something, please enlighten me.

sledgehammer999 replied:
>The explanation in its commit message seems still relevant. If I am missing something, please enlighten me.

I was quick to talk. Previously the UI responded immediately for every change during startup. Now it responds after the torrents have been loaded. So there's no chance of "losing" stuff.

What I don't understand is this: Why the fix worked? AFAIK, the mainloop doesn't start before `Base::exec()`. So the signals from the session wouldn't be tried to be delivered. So the mainwindow would have done its connects before Base::exec(). And after that the signals would have been delivered.

glassez replied:
>Why the fix worked?

Okay, let's leave it alone for now...

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+
+class StartUpTorrentsManager : public QObject
+{
+    Q_OBJECT
```
Q_DISABLED_COPY macro is also needed.

sledgehammer999 replied:
ok

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+#include <QVector>
+
+
+class StartUpTorrentsManager : public QObject
```
The name is confusing. It just loads the resume data. So ResumeDataLoadingManager is more appropriate name.

sledgehammer999 replied:
You're right. I was in a "hurry" to get the code working and I didn't give much thought to naming. (here and the other places you mentioned)

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+#include <QObject>
+#include <QVector>
+
+
```
Redundant blank line.

sledgehammer999 replied:
ok

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+
+#pragma once
+
+#include "base/bittorrent/magneturi.h"
```
application headers should be the last.

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+    };
+
+    explicit StartUpTorrentsManager(const QString &resumeFolderPath);
+    void resumeData(QVector<TorrentResumeData> &out);
```
getResumeData or getPendingResumeData (since it's not property).

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+    void readyRead();
+    void finished(int loadedCount);
+
+public slots:
```
Public slots should be before signals.

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+    void resumeData(QVector<TorrentResumeData> &out);
+
+signals:
+    void readyRead();
```
I would name it resumeDataReady.

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.h:
```diff
@@ -0,0 +1,72 @@
+
+signals:
+    void readyRead();
+    void finished(int loadedCount);
```
Thread can be finished, but Manager can't. It just can finish its job. So it should be something like loadingFinished.

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,244 @@
+#include "base/bittorrent/session.h"
+#include "base/utils/fs.h"
+
+#include <QDebug>
```
Wrong header order.

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,244 @@
+    int loadedCount = 0;
+    int nextQueuePosition = 1;
+    int numOfRemappedFiles = 0;
+    const QRegularExpression rx(QLatin1String("^([A-Fa-f0-9]{40})\\.fastresume$"));
```
AFAIK, QRegularExpression constructor accepts QLatin1String so you don't have to use it explicitly.

sledgehammer999 replied:
This is just copied from previous code, we was from a PR Chocobo1 merged a few hours ago...
But, if I don't explicitly pass a QLatin1String string will the raw string be accepted as QLatin1String or QString?

glassez replied:
Oops... forgive. There is no such constructor in this class. I confused it with some other class I used recently.

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,244 @@
+                    }
+                }
+
+                if (!m_signaledReady) {
```
I would use m_resumeDataReady name.

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,244 @@
+
+    if (numOfRemappedFiles > 0) {
+        logger->addMessage(
+            QString(tr("Queue positions were corrected in %1 resume files")).arg(numOfRemappedFiles),
```
Why QString cast here? tr returns QString.

sledgehammer999 replied:
old code, will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,244 @@
+    // readyresumeData might not have been emptied by the main thread yet
+    // so we need to append to that
+    QMutexLocker locker(&m_lock);
+    auto begin = queuedResumeData.cbegin();
```
begin confuses. Why not "it"?

sledgehammer999 replied:
will fix

glassez commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,244 @@
+        readyResumeData.push_back(begin.value());
+    queuedResumeData.clear();
+
+    if (!m_signaledReady) {
```
Shouldn't you check for readyResumeData.count() here?

sledgehammer999 replied:
Maybe we have signaled it already but the other thread hasn't taken the data yet. So no need to signal it again.
IMO, readyResumeData.count() can't indicate that state

glassez replied:
>IMO, readyResumeData.count() can't indicate that state

I mean `(readyResumeData.count() > 0) && !m_signaledReady`.
Otherwise it can emit signal in case when no data available.

glassez replied:
Or you can emit it in the loop above.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -537,14 +487,18 @@ Session::Session(QObject *parent)
     m_ioThread->start();
-    m_resumeDataTimer->start();
+
+    m_startupThread = new QThread(this);
```
You can use existing m_ioThread for this purpose. Why is yet another thread needed?

wesleybl commented:
@sledgehammer999 I did the following test with https://builds.shiki.hu/temp/qbittorrent_4.2.0alpha_x32_PR_9007.7z:

```
1. I downloaded a torrent completely to an external disk.
2. I stopped the qbittorrent.
3. I Removed the external disk.
4. I started qbittorrent.
5. The qbittorrent stopped the torrent, indicated "Files unavailable" and the download percentage was 0%.
```
In this situation I do not think the percentage should be 0%. As the disk is not there, qbittorrent does not know if I deleted the file.

If I connect the disk without stopping the qbittorrent and clicking "Resume", it starts downloading the file from scratch. But the file is already there.

If I stop the qbittorrent before connecting the disk, it recognizes that the file is there and maintains 100%.

Another test I did:

```
1. I downloaded a torrent completely to an external disk.
2. I deleted the files without stopping qbittorrent.
3. It keep 100% but every time someone tries to download the file from me, it indicates an i/o error. Indicate this several times.
```
It doesn't download the file again. What a breakthrough. But I should indicate I/O error only once.

Another:

```
1. I downloaded a torrent completely to an external disk.
2. I stopped the qbittorrent.
3. I deleted the files.
4. I started qbittorrent.
5. The qbittorrent stopped the torrent, indicated "Files unavailable" and the download percentage was 0%.
```
This is expected behavior. Awesome!

Great work! Thank you!

sledgehammer999 commented:
The 1st and 3rd test are essentially the same. In the point of view of the program, files are unavailable. Libtorrent doesn't differentiate between disk not reachable or files not there.
The progress stays 0% because at that point the "torrent job" is in an undetermined/undefined state. It needs both the files and the fastresume to indicate the progress.

The 2nd test is again limited by what libtorrent does. Libtorrent in the same session thinks that torrents are there even if they were deleted. Any failure is interpreted as a read error that might go away after a while.

wesleybl commented:
@sledgehammer999 I did the first test because I usually remove the external disk to copy files from it to another place, but I want to start qbittorrent to seeding files from other disks in the meantime. When I connect the external disk again, I want to continue seeding the torrents that are in it without having to check the files again (imagine that there are many torrents on the external disk).I think the qbittorrent would ideally indicate that the files are not there but keep 100%. So when I connect the disk I would only need to start the torrent with error, without it checking all. Anyway, I can solve this problem by stopping the qbitorrent before connecting the disk.

In the 2nd test the qbittorrent could be stop the torrent the first time it detected an I/O error.

sledgehammer999 commented:
>When I connect the external disk again, I want to continue seeding the torrents that are in it without having to check the files again (imagine that there are many torrents on the external disk).I think the qbittorrent would ideally indicate that the files are not there but keep 100%. So when I connect the disk I would only need to start the torrent with error, without it checking all.

We may be able to implement a workaround. When user presses "resume", we could  (behind the scenes) remove and re-add the torrent job so libtorrent can check again the fastresume and the data on disk. The only downside to this is that the torrent will probably lose its queue number.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -1349,11 +1308,7 @@ void Session::configure(libtorrent::settings_pack &settingsPack)
-    // Ignore -1 values because we don't want to set a max int message queue
-    settingsPack.set_int(libt::settings_pack::alert_queue_size, std::max(1000,
-        10 * std::max(maxActiveTorrents() * 2, maxConnections())));
+    settingsPack.set_int(libt::settings_pack::alert_queue_size, 1000000);
```
You need to declare named constant for this value.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -4114,8 +4021,9 @@ void Session::createTorrentHandle(const libt::torrent_handle &nativeHandle)
+        emit torrentAdded(torrent);
     // Send new torrent signal
     if (!data.resumed)
         emit torrentNew(torrent);
```
As I said before this additional signal isn't needed anymore.

wesleybl commented:
>When user presses "resume", we could (behind the scenes) remove and re-add the torrent job so libtorrent can check again the fastresume and the data on disk.

What is fastresume? How long does it take for each torrent? How will this influence deleted files? For example:

```
1. I downloaded two torrents to an external disk.
2. I stopped the qbittorrent.
3. I Removed the external disk.
4. I started qbittorrent.
5. The qbittorrent stopped the torrents, indicated "Files unavailable" and the download percentage was 0%.
6. On another computer I delete files from one of the torrents.
7. I reconnect the external drive.
8. I select the two torrents and click on resume.
```
What will happen to the torrent that had the files deleted? Ideally, it should go to and state "stopped" and do not start the download.

>The only downside to this is that the torrent will probably lose its queue number.

This would be an unusual situation. So I do not care if the queue number  is lost.

coolio2013 commented:
About https://builds.shiki.hu/temp/qbittorrent_4.2.0alpha_x32_PR_9007.7z on Win XP-32:

Please rethink behaviour to disable the GUI. Rechecking takes time, especially big torrents on network shares.

Torrents with previous status 'Paused' (set manually to avoid recheck) and dirty from previous crashes WILL be rechecked at startup of qbt. Nothing you can do, since the GUI is all disabled. LOG says 

`Fast resume data was rejected for torrent {payload} fast resume rejected. check resume(): missing or invalid 'file sizes' entry. Checking again...`

process-monitor reveals that qbt does communications with other peers (I can't see checking itself). 
fastresume-files are not rewritten (same time/date for over 8hours) after start; some torrents should have completed recheck meanwhile. So the status remains and they will be rechecked again at next start, if qbt crashes or if I kill it now.

This will repeat forever. qbt will become completely useless. Please consider that.

harrv commented:
I'd really like to see this PR merged. When I use the file system to delete or move a file that I had been seeding, I never want it to be re-downloaded. I'd prefer to see an error in the client and then decide if I want to remove the affected torrent from the client or change that torrent's file path. I got used to the beauty and simplicity of that workflow when I used uTorrent.

Is there anything I can do to help with the PR merge process? @sledgehammer999 thank you very much for your code changes to resolve this. Have you seen the review from @glassez on June 2? The requested changes look minor. Is that the only thing holding up merging this PR?

glassez commented:
>The requested changes look minor. Is that the only thing holding up merging this PR?

The requested changes are prerequisites for the main review.

glassez commented:
>Since this is an interactive program, displaying no GUI would be a bad choice.
If feasible, I would choose to ready the mainwindow as early as possible (modern browsers startup time is impressive!) and let user know program is still initializing: could be some text in statusbar, or display the torrent jobs initially as "not yet loaded" and transition to other ready state (paused or downloading) when it is loaded.

I've been thinking a lot about this problem lately. And I've come to the conclusion again that those disruptions to the natural sequence of initialization of the components with that you guys are trying to solve this problem can't lead to anything good. We will only get a lot of logical errors and unmaintainable code!
The sequence should be in used-using order (as I said many times before). At global it should be as the following:
1. Core components
2. UI and Remote Access components

But @Chocobo1 is right too, the qBittorrent is "interactive program" (in GUI mode) so it would be incorrect to leave the user without a response for a long time. However, this does not mean that we must break the right order. Why should we display the main window immediately if it can't do any good at this time?..
I suggest another option (similar to other long-loaded programs): we just show the splash window (with some initialization related messages) until the application is fully loaded. Tray icon can also be showed but it shouldn't provide any controls during application loading.

tehmasterer commented:
Ugh... I thought this was fixed with the 4.1.1 update but it wasn't. I launched qbittorrent last week and I got all these i/o error pop notifications because my external drive wasn't connected. I noticed it switched all those torrents to resume so I quickly paused them. Today I connected my drive and a few of my completed, paused torrents, started overwriting.

If a torrent is completed but it can't find it, can it just say 100% and STAY paused? The overwriting is an absolute nightmare. Also some torrents have to be rechecked after I connect my external drive, even though the save path is the same. This is annoying for seeding.

glassez commented:
@sledgehammer999, I have an idea (I need to test it). What about restoring completed torrents in "seeding mode"? This should prevent it from being rechecked. But I need to test it in different use cases (e.g. when torrent was downloaded partially and then user want to download some more files).

sledgehammer999 commented:
FYI, I'll be able to resume this PR either today or the weekend.

> What about restoring completed torrents in "seeding mode"?

This sounds like a solution too. But you'll need to inspect the resume data yourself to determine the status.
Plus, you need to check for paused torrent instead (non-completed ones can have the same problem).

sledgehammer999 commented:
PR rebased and review comments addressed.
>
>```c++
> if (!data.resumed)
>         emit torrentNew(torrent);
>```
>As I said before this additional signal isn't needed anymore.

This signal is only used in the mainwindow in order to display a tray notification for **newly** added torrents.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -4351,6 +4262,39 @@ void Session::handleExternalIPAlert(libt::external_ip_alert *p)
 
+void Session::handleStartUpFinished()
+{
+    m_startupInProgress = false;
```
This flag is confusing. It splits lifetime of Session in 3 steps (uninitialized, starting, started), but we can handle only last two.
It's better to use m_started flag (initially false and assigning true here).

sledgehammer999 replied:
OK. That seems to work too.

glassez commented on src/app/application.cpp:
```diff
@@ -498,53 +498,19 @@ int Application::exec(const QStringList &params)
-#endif // DISABLE_WEBUI
-#else
+#ifndef DISABLE_GUI
     m_window = new MainWindow;
```
Here you have some subsystems related GUI (RSS, Search) that assume the handled subsystems are already initialized, but you still didn't do it. Or I'm wrong?

sledgehammer999 replied:
Look at `MainWindow::setEnabledWidgets()`. I think I handle all the cases.

glassez replied:
You missed SearchWidget there.

sledgehammer999 commented on src/app/application.cpp:
```diff
@@ -498,53 +498,19 @@ int Application::exec(const QStringList &params)
 
-    m_running = true;
+    // Slots are executed in the order of connection.
+    // Thus we need this connected last.
```
Good question. Honestly, I can't remember why I wrote this.
It is probably a leftover of an earlier version before I published the PR.
I'll delete the comment.

sledgehammer999 replied:
I think I know why I have this here.
You need to look the previous commit https://github.com/qbittorrent/qBittorrent/pull/9007/commits/bc8e4f8e946a170bb1803012133150c0685ad23c
There some GUI widgets connect to the same signal in order to initialize themselves. eg the transferlist.
We want them to do it first while they're disabled and then enable them

glassez replied:
I don't think this order is so important, because all these processes are taking place in a single event...

sledgehammer999 commented on src/base/bittorrent/private/startuptorrentsmanager.cpp:
```diff
@@ -0,0 +1,237 @@
+ * exception statement from your version.
+ */
+
+#include "startuptorrentsmanager.h"
```
ok

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -531,6 +481,11 @@ Session::Session(QObject *parent)
+
+    m_resumeDataLoadingManager = new ResumeDataLoadingManager(m_resumeFolderPath);
+    m_resumeDataLoadingManager->moveToThread(m_ioThread);
+    connect(m_ioThread, &QThread::finished, m_resumeDataLoadingManager, &QObject::deleteLater);
```
You need also add:
```c++
connect(m_resumeDataLoadingManager, &ResumeDataLoadingManager::loadingFinished, m_resumeDataLoadingManager, &QObject::deleteLater);
```

Also I would move its initialization it `startUpTorrents()` and make it local variable since it isn't needed anywhere else.

glassez commented on src/gui/mainwindow.h:
```diff
@@ -258,6 +260,7 @@ private slots:
 #endif
     bool m_hasPython;
     QMenu *m_toolbarMenu;
+    bool m_sessionStarted;
```
I would add `Session::isStarted() const { return m_started; }` getter and use it where needed instead of having separate variables.

glassez commented on src/gui/mainwindow.h:
```diff
@@ -100,6 +100,8 @@ class MainWindow : public QMainWindow
 
     void showNotificationBaloon(QString title, QString msg) const;
+    // After Application has handled BitTorrent::Session::startupFinished
+    void setEnabledWidgets(bool enabled);
```
It's better to declare it private and connect to `Session::startupFinished()` signal somewhere in MainWindow constructor.

glassez commented on src/gui/mainwindow.cpp:
```diff
@@ -1302,8 +1308,11 @@ void MainWindow::dropEvent(QDropEvent *event)
 {
-    foreach (const QString &mime, event->mimeData()->formats())
-        qDebug("mimeData: %s", mime.toLocal8Bit().data());
+    if (!m_sessionStarted) {
```
You also need to prevent torrent addition in `Session::addTorrent()`.

sledgehammer999 commented:
@glassez I did a minor PR update.

Because the modules are so interconnected we need to serialize some actions in the startup.
Until the torrents are loaded, new torrents can't be added. This also preserves the order of the loaded torrents.
So RSS and WebUI are started when the torrents are loaded. The MainWindow disables all other entry points (include the search tab).
I investigated if there was an easy way to do it inside `addTorrent_impl()` but I couldn't.

This also affects the order we connect to `startupFinished` signal. The Application needs to be last to react. It is the one that initializes RSS, WebUI and enables all the elements of MainWindow. The order of current code is preserved.
Also, theoretically, widgets draw their contents quicker if they have received **all** their data while they were hidden. (eg transferlist with hundreds of torrent entries)
That's why I call `MainWindow::setEnabledWidgets()` from `Application::handleStartUpFinished()`.

`m_sessionStarted` is a guard inside `MainWindow::setEnabledWidgets()` and I can't think of a way to eliminate it and make use of `Session::isStarted()`.

glassez commented:
@sledgehammer999, unfortunately I can't say that I approve this PR in its current form. It's more like a rough workaround (in terms of its design).
However, as I see it, it takes away from you the resources that are already missing for full maintainence of the project. So you are free to merge it if you think its purpose is achieved (from a user point of view, closing our eyes to its implementation) in the hope that we will polish and improve it in the future.
But I would make upcoming release ASAP (without this PR) and then merge it immediately so we would be able to test it thoroughly and fix/improve it before the next release.

P.S. Have you ever thought of having an additional release manager? This could help to make releases on time. It is a very hateful situation, when hotfixes "rot" in the master for several months while the bugtracker is filled with duplicate issues!

glassez commented:
@sledgehammer999, you still want to do something about it?

glassez commented:
@sledgehammer999, https://github.com/qbittorrent/qBittorrent/pull/9007#issuecomment-450510074

githubuserthatsright commented:
Just so that I understand - this potential fix, or for that matter maybe any other, has been around for a bit over a year and more than just a few version updates and we've still got this as a problem?

glassez commented:
>Just so that I understand - this potential fix, or for that matter maybe any other, has been around for a bit over a year and more than just a few version updates and we've still got this as a problem?

The problems it was supposed to solve (at least most of them) are now solved in a different way.

githubuserthatsright commented:
So moving or deleting data of a formerly completed torrrent will in no case any longer cause a redownload?

glassez commented:
>So moving or deleting data of a formerly completed torrrent will in no case any longer cause a redownload?

This should be fixed now.

glassez commented:
The goal was achieved in a different way.
